### PR TITLE
[WIP] lang, ts: idl versioning

### PR DESCRIPTION
**Pull request by armaniferrante**
Introduces a new header to accounts, instructions, and events:

```
<1-byte-layout-version> || <1-byte-type-version> || <4-byte-discriminator> || <2-bytes-reserved>
```

For existing anchor programs, one should use the `deprecated-layout` feature in one's Cargo.toml to maintain backwards compatibility with the existing 8 byte discriminator. Clients do not have to be updated and will be backwards compatible automatically, since they will use the IDL to infer which layout to use.

TODO:

- add layout version to account, instruction, and event 
- instructions test
- events test
- revert https://github.com/project-serum/anchor/pull/1490 before merging (this pr makes the close trait safe to use outside the macros)

### CODE REVIEW

File: lang/attribute/account/src/lib.rs [hunk 1]
```diff
@@ -137,20 +142,19 @@ pub fn account(
...
                 // This trait is useful for clients deserializing accounts.
                 // It's expected on-chain programs deserialize via zero-copy.
                 #[automatically_derived]
                 impl #impl_gen anchor_lang::AccountDeserialize for #account_name #type_gen #where_clause {
                     fn try_deserialize(buf: &mut &[u8]) -> std::result::Result<Self, ProgramError> {
-                        if buf.len() < #discriminator.len() {
-                            return Err(anchor_lang::error::ErrorCode::AccountDiscriminatorNotFound.into());
+                        // Header is always 8 bytes.
+                        if buf.len() < anchor_lang::accounts::header::HEADER_LEN {
+                            return Err(anchor_lang::__private::ErrorCode::AccountDiscriminatorNotFound.into());
```
Comment (armaniferrante) [inline lang/attribute/account/src/lib.rs:155]:
```suggestion
                            return Err(anchor_lang::error::ErrorCode::AccountDiscriminatorNotFound.into());
```

**Comment by paul-schaaf** [comment]
this will close #233

**Comment by LordOfTheDings** [comment]
Is there an estimation as to when this is going to be shipped? Thx!

**Comment by armaniferrante** [comment]
Closing as stale.

**Pull request closed without merging**